### PR TITLE
Fix halfCommitter and WithTx

### PR DESCRIPTION
**Pull request by wolfogre**
Related to #22362.

I overlooked that there's always `committer.Close()`, like:

```go
		ctx, committer, err := db.TxContext(db.DefaultContext)
		if err != nil {
			return nil
		}
		defer committer.Close()

		// ...

		if err != nil {
			return nil
		}

		// ...

		return committer.Commit()
```

So the `Close` of `halfCommitter` should ignore `commit and close`, it's not a rollback.

See: [Why `halfCommitter` and `WithTx` should rollback IMMEDIATELY or commit LATER](https://github.com/go-gitea/gitea/pull/22366#issuecomment-1374778612).

### CODE REVIEW

**Comment by lunny** [comment]
If it's in a database transaction, `committer` should do nothing, don't commit and don't close.

**Comment by wolfogre** [comment]
> If it's in a database transaction, `committer` should do nothing, don't commit and don't close.

Why `halfCommitter` shouldn't ignore `rollback and close`:

```go
func A(ctx context.Context) error {
	ctx, committer, err := db.TxContext(ctx)
	if err != nil {
		return nil
	}
	defer committer.Close()

	// insert something
	
	if err := B(ctx); err != nil {
		return err
	}

	return committer.Commit()
}

func B(ctx context.Context) error {
	ctx, committer, err := db.TxContext(ctx)
	if err != nil {
		return nil
	}
	defer committer.Close()

	// insert something
	
	// check something
	if err != nil {
		log.Error("it's ok to ignore this error and give up all data")

		// it's ok to ignore this error, but should rollback the data inserted previously
		return nil
	}

	// insert something

	return committer.Commit()
}
```

The `B` function tries to be quiet, neither returning an error nor inserting any data. However, if the `Close` of `halfCommitter` does nothing when rollbacking, some incomplete data will be written.

**Comment by wolfogre** [comment]
BTW, I think the same problem is going to happen with `WithTx`. So shall we rollback immediately when an error occurs? Like:

```diff
func WithTx(parentCtx context.Context, f func(ctx context.Context) error) error {
	if sess, ok := inTransaction(parentCtx); ok {
-		return f(newContext(parentCtx, sess, true))
+		err := f(newContext(parentCtx, sess, true))
+		if err != nil {
+			// let any commits later do nothing
+			_ = sess.Close()
+		}
+		return err
	}
	return txWithNoCheck(parentCtx, f)
}
```

I'm not quite sure now, or do we agree that a function always needs to return an error after rollback a transaction?

I need more advice.

**Review by lunny — APPROVED** [review]

File: models/db/context_committer_test.go [hunk 1]
```diff
@@ -0,0 +1,102 @@
...
+		if err != nil {
+			return nil
+		}
+
+		// ...
+
+		return committer.Commit()
+	*/
+
+	testWithCommiter := func(committer Committer, f func(committer Committer) error) {
```
Comment (delvh) [inline models/db/context_committer_test.go]: typo.

**Review by delvh — APPROVED** [review]

**Pull request merged**